z3/src/z3_factory.cpp

```diff
@@ -1,15 +1,48 @@
+/*********************                                                        */
+/*! \file z3_factory.cpp
+** \verbatim
+** Top contributors (to current version):
+**   Amalee Wilson
+** This file is part of the smt-switch project.
+** Copyright (c) 2020 by the authors listed in the file AUTHORS
+** in the top-level source directory) and their institutional affiliations.
+** All rights reserved.  See the file LICENSE in the top-level source
+** directory for licensing information.\endverbatim
+**
+** \brief Factory for creating a Z3 SmtSolver
+**
+**
+**/
+
 #include "z3_factory.h"
 
 #include "z3_solver.h"
 
-namespace smt
-{
+#include "logging_solver.h"
+
+namespace smt {
 
-/* Z3SolverFactory implementation */
-SmtSolver Z3SolverFactory::create()
+bool initialized = false;
+
+/* Z3SolverFactory implementation with logging */
+SmtSolver Z3SolverFactory::create(bool logging)
 {
-  return std::make_unique<Z3Solver>();
+  // Must initialize only once.
+  // Different instances of the solver will have
+  // different contexts.
+  // if (!initialized)
+  // {
+  //   yices_init();
+  //   initialized = true;
+  // }
+
+  SmtSolver solver = std::make_shared<Z3Solver>();
+  // if (logging)
+  // {
+  //   solver = std::make_shared<LoggingSolver>(solver);
+  // }
+  return solver;
 }
-/* end Z3SolverFactory implementation */
+/* end Z3SolverFactory logging implementation */
```

Comment:
I don't think you need the "logging", but I don't feel strongly either way. The factory supports both with and without logging (btw, logging just keeps track of the term DAG more accurately at the smt-switch level instead of relying on the solver).